core/src/main/scala/quasar/plugin/jdbc/datasource/JdbcDatasource.scala

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2020 Precog Data
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.plugin.jdbc.datasource
+
+import quasar.plugin.jdbc._
+import quasar.plugin.jdbc.implicits._
+
+import scala.{Boolean, None, Option, Some}
+import scala.util.{Left, Right}
+
+import java.lang.Throwable
+
+import cats.Defer
+import cats.data.{Ior, NonEmptyList}
+import cats.effect.{Bracket, Resource}
+import cats.implicits._
+
+import doobie._
+import doobie.implicits._
+
+import fs2.{Pull, Stream}
+
+import quasar.api.resource.{ResourceName, ResourcePath, ResourcePathType => RPT}
+import quasar.api.datasource.DatasourceType
+import quasar.connector.QueryResult
+import quasar.connector.datasource.{LightweightDatasourceModule, Loader}
+import quasar.qscript.InterpretedRead
+
+import shims.equalToCats
+
+final class JdbcDatasource[F[_]: Bracket[?[_], Throwable]: Defer] private (
+    xa: Transactor[F],
+    discovery: JdbcDiscovery,
+    val kind: DatasourceType,
+    val loaders: NonEmptyList[Loader[Resource[F, ?], InterpretedRead[ResourcePath], QueryResult[F]]])
+    extends LightweightDatasourceModule.DS[F] {
+
+  def pathIsResource(path: ResourcePath): Resource[F, Boolean] =
+    resourcePathRef(path).fold(false.pure[Resource[F, ?]]) {
+      case Left(table) =>
+        Resource.liftF(discovery.tableExists(table, None).transact(xa))
+
+      case Right((schema, table)) =>
+        Resource.liftF(discovery.tableExists(table, Some(schema)).transact(xa))
+    }
+
+  def prefixedChildPaths(prefixPath: ResourcePath): Resource[F, Option[Stream[F, (ResourceName, RPT.Physical)]]] = {
+    type Out[X[_]] = Stream[X, (ResourceName, RPT.Physical)]
+
+    if (prefixPath === ResourcePath.Root)
+      discovery.topLevel
+        .map(_.fold(
+          s => (ResourceName(s.asString), RPT.prefix),
+          t => (ResourceName(t.asString), RPT.leafResource)))
+        .transact(xa)
+        .some
+        .pure[Resource[F, ?]]
+    else
+      resourcePathRef(prefixPath).fold((None: Option[Out[F]]).pure[Resource[F, ?]]) {
+        case Right((schema, table)) =>
+          Resource liftF {
+            discovery.tableExists(table, Some(schema))
+              .map(p => if (p) Some(Stream.empty: Out[F]) else None)
+              .transact(xa)
+          }
+
+        case Left(ident) =>
+          def paths =
+            discovery.tables(Ior.left(ident))
```

Comment:
Here, the `Left(ident)` uses `ident` as a schema identifier. But up on line 54 in this file, `Left(table)` uses `table` as a table identifier. Shouldn't `resourcePathRef` behave consistently?

Reply:
Do we want `Ior.right(ident)` here instead?

Reply:
The larger question really is: when `resourcePathRef` returns a single value, (not a tuple), how do we know if that value is the schema name or the table name?

Reply:
I named it `ident` to convey the ambiguity, as you point out, we don't yet know if it refers to a schema or a table, it is just an identifier.

> The larger question really is: when resourcePathRef returns a single value, (not a tuple), how do we know if that value is the schema name or the table name?

Indeed, this is handled below, we first check if `ident` is a table without a schema, if so we're done, else we treat it as a schema and returns `paths`.

Reply:
Ohh I see how this is working now! `paths` above is only actually used when we land in the `!isTable` case below. All good! :)